### PR TITLE
Sensor Status Queue for Race Condition

### PR DESCRIPTION
**Pull request by BraveKerry**
Want to double check my implementation.

### CODE REVIEW

File: LoraBraveSensor.ino [hunk 1]
```diff
@@ -7,46 +7,49 @@
...
+    api.system.timer.create((RAK_TIMER_ID)FSM_TIMER, (RAK_TIMER_HANDLER)fsm::addToSensorDataQueue, RAK_TIMER_ONESHOT);
     api.system.timer.start((RAK_TIMER_ID)FSM_TIMER, sleepDuration, (void *)1);
   }
-  api.system.sleep.all();
+  if (fsm::stateQueueEmpty()) {
+    DEBUG_SERIAL_LOG.println("State queue empty, sleeping");
+    if (lora::isUplinkProcess()) {
+      delay(sleepDuration); 
+    } else {
+      api.system.sleep.all();
```
Comment (BraveKerry) [inline LoraBraveSensor.ino]: Potential race condition, if interrupt (`fsm::addToSensorDataQueue`) runs between the queue emptiness check at line 31 through to line 47?

File: fsm.cpp [hunk 2]
```diff
@@ -18,12 +20,29 @@ int state1_countdown_timer = 0;
 int state2_duration_timer = 0;
 int state3_stillness_timer = 0;
 
+std::queue<sensorData_t> sensorDataQueue;
```
Comment (trisapeace) [inline fsm.cpp]: I'm still trying to work through what I'd recommend, but take a look at this article on a threadsafe queue: https://codetrips.com/2020/07/26/modern-c-writing-a-thread-safe-queue/ . I think the person does a good job of articulating what needs to be done for sharing a queue across different threads.

Comment (trisapeace) [reply fsm.cpp]: What I'm not sure about is whether these interrupts run their code in a new thread or not.

File: fsm.cpp [hunk 3]
```diff
@@ -18,12 +20,29 @@ int state1_countdown_timer = 0;
...
+DoorSensor doorSensor = DoorSensor(DOOR_SENSOR_PIN);
+MotionSensor motionSensor = MotionSensor(MOTION_SENSOR_PIN);
```
Comment (trisapeace) [inline fsm.cpp]: There is a note in the Arduino docs saying that the `volatile` keyword should be used for variables shared by the interrupt handlers and the main code: https://www.arduino.cc/reference/en/language/functions/external-interrupts/attachinterrupt/ . So maybe all of these need it. However, all the examples that I see in Reply #7 of http://gammon.com.au/interrupts use primitive types like `int` and `bool`. Whereas the ones that you have are objects, which means that they should just be pointers which won't actually change. So it might not apply here. More to investigate.

File: LoraBraveSensor.ino [hunk 4]
```diff
@@ -7,46 +7,49 @@
...
   {
-    api.system.timer.create((RAK_TIMER_ID)FSM_TIMER, (RAK_TIMER_HANDLER)[](void *){}, RAK_TIMER_ONESHOT);
+    api.system.timer.create((RAK_TIMER_ID)FSM_TIMER, (RAK_TIMER_HANDLER)fsm::addToSensorDataQueue, RAK_TIMER_ONESHOT);
     api.system.timer.start((RAK_TIMER_ID)FSM_TIMER, sleepDuration, (void *)1);
   }
-  api.system.sleep.all();
+  if (fsm::stateQueueEmpty()) {
+    DEBUG_SERIAL_LOG.println("State queue empty, sleeping");
+    if (lora::isUplinkProcess()) {
+      delay(sleepDuration);
```
Comment (trisapeace) [inline LoraBraveSensor.ino]: Sounds like the use of the `delay` function is discouraged because the program won't do anything else while it's waiting: 

https://www.arduino.cc/reference/en/language/functions/time/delay/
http://gammon.com.au/interrupts

Why did you add this in  here now but it wasn't there before?

File: systemTimers.h [hunk 5]
```diff
@@ -4,5 +4,6 @@
 #include <Arduino.h>
 
 const unsigned int FSM_TIMER = RAK_TIMER_0; // Timer to tell how long system to sleep for during each state
+const unsigned int UPLINK_PROCESS_TIMER = RAK_TIMER_1; // Timer to change the uplink process to false after a certain amount of time
```
Comment (trisapeace) [inline systemTimers.h]: I was going to ask you whether this could be done instead of using `delay`, but now I see that you've commented this out elsewhere. Did it not work like you expected?

File: README.md [hunk 6]
```diff
@@ -56,6 +56,7 @@ decodes to `{"alertType":"DurationAlert","battery":8192,"countdownTimer":15,"dur
 
 ## Power Saving Measures
 The system is designed to go to sleep, unless there are changes to the digital input pins, or a timer interrupt. Limitation's of RUI3, results in a specific implementation -- more info can be found [here](https://forum.rakwireless.com/t/rui3-wake-on-interrupt/7460/8). 
+RAK_TIMER_#, are one of five hardware timers provided by RUI3.
```
Comment (trisapeace) [inline README.md:67]: Thanks!

File: fsm.cpp (hunk 3 above)
Comment (trisapeace) [reply fsm.cpp]: I'd lean towards adding the `volatile` keyword just in case. It might propagate down to the primitives that it's made up of like it sort of describes here about structs: https://en.cppreference.com/w/c/language/volatile

File: fsm.cpp [hunk 7]
```diff
@@ -18,12 +20,29 @@ int state1_countdown_timer = 0;
...
+int fsm::handleState()
+{
+    if (!sensorDataQueue.empty()) {
+        sensorData_t sensorData = sensorDataQueue.front();
+        sensorDataQueue.pop();
+        return stateHandler(sensorData);
+    }
+    sensorData_t sensorData = {
+        .doorOpen = doorSensor.isDoorOpen(),
+        .motion = motionSensor.isThereMotion()};
```
Comment (trisapeace) [inline fsm.cpp]: Why do you need to read this here instead of just getting it off the queue?

File: fsm.cpp [hunk 8]
```diff
@@ -18,12 +20,29 @@ int state1_countdown_timer = 0;
...
+    return stateHandler(sensorData); 
+}
```
Comment (trisapeace) [inline fsm.cpp]: I would find this easier to read if there were just one return statement. If you declare `sensorData_t sensorData` at the top of the function and use an if/else instead of just a single if, then the return can go at the bottom.  (Assuming you even need the `else` part.)

File: sensors.h [hunk 9]
```diff
@@ -91,6 +94,11 @@ class AnalogSensor
     int getValue();
 }; 
 
+struct sensorData_t {
+    bool doorOpen; // door open?
+    bool motion; // motion?
+};
```
Comment (trisapeace) [inline sensors.h]: To my eye, it seems weird to have a struct name start with a small letter and have the `_t` on the end because I think of structs as classes without functions. But I don't code in C++ very often so am very likely not used to C++ conventions. Is this something that you saw elsewhere?

File: lora.h [hunk 10]
```diff
@@ -51,5 +51,16 @@ namespace lora
      * the callback function for the LoRa module to call when a message is sent
      */
     static void sendCallback(int32_t status);
+
+    /**
+     * @returns whether or not a sending process is currently occuring or not
+     */
+    bool isUplinkProcess();
```
Comment (trisapeace) [inline lora.h]: What do you think about renaming this `isUplinkInProgress`?

File: lora.cpp [hunk 11]
```diff
@@ -227,4 +222,25 @@ void lora::sendUplink(lora::uplinkMessage msg)
...
+
+    if (loraTimer <= 0)
+    {
+        uplinkProcess = false; 
+        DEBUG_SERIAL_LOG.println("Uplink routine complete");
+        return INT_MAX; 
+    }
+    return loraTimer;    
+  }
+  return INT_MAX;
```
Comment (trisapeace) [inline lora.cpp]: Can you please explain how this function fits into things? I'm not sure that I understand what it's for. Thanks!

File: fsm.h [hunk 12]
```diff
@@ -49,20 +49,37 @@ namespace fsm
...
+    /**
+     * adds to the fsm queue the next dataset for the FSM to process
+     */
+    void addToSensorDataQueue(); 
+
+    /**
+     * check if the fsm queue is empty
+     * @return true if empty, false otherwise
+     */
+    bool stateQueueEmpty();
```
Comment (trisapeace) [inline fsm.h]: Are both of these functions talking about the same queue? If so, what do you think about making the names more consistent?
`addToSensorDataQueue` and `isSensorDataQueueEmpty`
or 
`addToStateQueue` and `isStateQueueEmpty`

File: heartbeat.cpp [hunk 13]
```diff
@@ -10,7 +10,7 @@ static int HEARTBEAT_INTERVAL = 300000;
 static int heartbeatTimer = 0; 
 static int lastHeartbeatHandleTime = 0;
 
-unsigned int heartbeat::getRemainingDuration()
```
Comment (trisapeace) [inline heartbeat.cpp:17]: Just curious, why did these change from `unsigned int` to `int`?

File: lora.cpp [hunk 14]
```diff
@@ -1,19 +1,22 @@
...
-bool sending = false;
-bool receiving = false;
+const int DOWNLINK_INTERVAL = 5000; 
+int lastHandledTime = millis(); 
+int loraTimer = DOWNLINK_INTERVAL; 
+bool uplinkProcess = false;
 
 static void lora::recvCallback(SERVICE_LORA_RECEIVE_T *data)
 {
-  receiving = true;
```
Comment (trisapeace) [inline lora.cpp:16]: You no longer need to track when you're receiving a downlink?

File: lora.cpp [hunk 15]
```diff
@@ -61,7 +63,10 @@ static void lora::sendCallback(int32_t status)
   {
     DEBUG_SERIAL_LOG.println("Sending failed (callback)");
   }
-  sending = false;
+  DEBUG_SERIAL_LOG.println("Lora send complete, checking for downlinks"); 
+  // api.system.timer.create((RAK_TIMER_ID)UPLINK_PROCESS_TIMER, (RAK_TIMER_HANDLER)[](void *) {uplinkProcess = false; 
+  // DEBUG_SERIAL_LOG.println("End of uplink routine"); }, RAK_TIMER_ONESHOT);
+  // api.system.timer.start((RAK_TIMER_ID)UPLINK_PROCESS_TIMER, 5000, (void *)1);
```
Comment (trisapeace) [inline lora.cpp]: Normally, I recommend taking out any commented-out code. Is there a particular reason to leave this here?

File: lora.cpp [hunk 16]
```diff
@@ -186,19 +194,6 @@ void lora::sendUplink(char *payload)
       DEBUG_SERIAL_LOG.printf("Sending request attempt %i\r\n", i);
```
Comment (trisapeace) [inline lora.cpp]: As mentioned in a different comment, it seems that they discourage the use of `delay`. Is there another way to do this that is still energy-efficient enough for the battery-powered version?

File: sensors.h [hunk 17]
```diff
@@ -91,6 +94,11 @@ class AnalogSensor
     int getValue();
 }; 
 
+struct sensorData_t {
+    bool doorOpen; // door open?
+    bool motion; // motion?
```
Comment (trisapeace) [inline sensors.h]: I liked what you had been doing when you were naming `bool` values starting with the word `is`. So `isDoorOpen` and `isMotion` (or `isThereMotion` to be consistent with the function name) to me was really nice.

**Review by trisapeace — COMMENTED** [review]
Thanks for implementing the queue. I think there are still some issues (see my comments), but this is a good start. Asynchronous code is a difficult thing to do correctly, but it's important. So I appreciate you learning about this and taking the time to do things right for what it's worth at this point in the development.

@sampathsatti how close to production-ready do you want this to be at this stage of the game?

File: fsm.cpp (hunk 2 above)
Comment (BraveKerry) [reply fsm.cpp]: I'll use their thread-safe queue then.

File: fsm.cpp (hunk 3 above)
Comment (BraveKerry) [reply fsm.cpp]: I believe the only variable used by interrupt handlers is the queue, therefore using `volatile` would be unnecessary.  Unless you believe modifying the thread-safe queue with the `volatile` keywords, can improve it.

File: fsm.cpp (hunk 7 above)
Comment (BraveKerry) [reply fsm.cpp]: With the current implementation, `fsm::handleState()` can be called while the queue is empty.

File: sensors.h (hunk 9 above)
Comment (BraveKerry) [reply sensors.h]: I'll change it to `SensorData`. I was more or less thinking it was a data type rather than a class, but I can see how it would make sense to name it as if it were a class.

File: lora.cpp (hunk 11 above)
Comment (BraveKerry) [reply lora.cpp]: This is for keeping track the amount of time that has passed since the last uplink. So that once the given time has passed, I can put the cpu/lora to sleep (as putting them to sleep during a uplink causes them to be continuously woken up, presumably from RUI3).

File: heartbeat.cpp (hunk 13 above)
Comment (BraveKerry) [reply heartbeat.cpp:17]: This is a bugfix, as the value of heartbeatTimer can be negative from `heartbeat::getRemainingDuration()`'s first line.

File: lora.cpp (hunk 14 above)
Comment (BraveKerry) [reply lora.cpp:16]: I grouped the uplink and downlink wait time together to the `uplinkProcess` boolean.

File: lora.cpp (hunk 15 above)
Comment (BraveKerry) [reply lora.cpp]: Sorry, this is my old uplink timer code. I'll probably end up going back to it. I just wanted for you to check out my queue implementation asap, as I have no idea if what I was doing made sense.

File: lora.cpp (hunk 16 above)
Comment (BraveKerry) [reply lora.cpp]: Let me get back to you on this. From current testing, it feel like if it fails the first time, it'll fail the second time too, so I may just remove this part entirely. If not, I can definitely implement method without the delays.

File: fsm.cpp (hunk 7 above)
Comment (BraveKerry) [reply fsm.cpp]: On second thought, this is fixable.

File: fsm.cpp (hunk 2 above)
Comment (BraveKerry) [reply fsm.cpp]: As discussed, I was unable to. I have added a note.

File: LoraBraveSensor.ino (hunk 4 above)
Comment (BraveKerry) [reply LoraBraveSensor.ino]: Previously, I had the blocking delays in the lora code. I have reverted it to that, as I believe it improves uplink consistency.

File: systemTimers.h (hunk 5 above)
Comment (BraveKerry) [reply systemTimers.h]: I have tried using a timer to avoid using `delay`, but this significantly reduces uplink success consistency.

File: lora.cpp (hunk 15 above)
Comment (BraveKerry) [reply lora.cpp]: I have now removed this code, as the `delay` method improves uplink consistency.

File: lora.cpp (hunk 16 above)
Comment (BraveKerry) [reply lora.cpp]: I was correct. I have removed the loop altogether.

File: fsm.cpp (hunk 3 above)
Comment (trisapeace) [reply fsm.cpp]: Removed `volatile`s because they are only on classes and this will not auto-magically make them threadsafe, unfortunately.

File: LoraBraveSensor.ino (hunk 4 above)
Comment (trisapeace) [reply LoraBraveSensor.ino]: We're unsure why this is advised against. So we're still using it.

**Review by trisapeace — APPROVED** [review]
Thanks for going through this with me, Kerry. If you're feeling confident with your testing, then I'm happy for this to be "Squash and Merge"-ed. Congrats on all your success getting this up and running!
